styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java

```diff
@@ -91,72 +104,110 @@ void tick() {
-            .event(workflow, Trigger.backfill(backfill.id()), partition)
-            .toCompletableFuture();
-        // Wait for the trigger execution to complete before proceeding to the next partition
-        processed.get();
-      } catch (AlreadyInitializedException ignored) {
-        // nop
-      } catch (Throwable e) {
-        LOG.warn("failed to trigger backfill for state [{}]: {}", partition, backfill, e);
-        return;
+        if (!storage.runInTransaction(tx ->
+            triggerNextPartitionAndProgress(tx, backfill.id(), workflow,
+                initialNextTrigger, remainingCapacity))) {
+          break;
+        }
+      } catch (IOException e) {
+        // transaction failure, yield
+        LOG.debug("Transaction failure when working on backfill {}", backfill, e);
+        break;
       }
+    }
+  }
 
-      partition = nextInstant(partition, backfill.schedule());
-      storeBackfill(backfill.builder()
-                        .nextTrigger(partition)
-                        .build());
+  private boolean triggerNextPartitionAndProgress(StorageTransaction tx,
+                                                  String id,
+                                                  Workflow workflow,
+                                                  Instant initialNextTrigger,
+                                                  int remainingCapacity) {
+    final Backfill momentBackfill = tx.backfill(id).orElseThrow(RuntimeException::new);
+    final Instant momentNextTrigger = momentBackfill.nextTrigger();
+
+    if (numberOfInstants(momentNextTrigger, initialNextTrigger,
+        momentBackfill.schedule()) >= remainingCapacity) {
+      LOG.debug("Capacity reached for backfill {}", momentBackfill);
+      return false;
     }
 
-    if (partition.equals(backfill.end())) {
-      storeBackfill(backfill.builder()
-                        .nextTrigger(backfill.end())
-                        .allTriggered(true)
-                        .build());
+    if (momentNextTrigger.equals(momentBackfill.end())) {
+      LOG.debug("Backfill {} all triggered", momentBackfill);
+      return false;
     }
-  }
 
-  private Map<String, Long> getBackfillStates() {
-    final List<InstanceState> activeStates = stateManager.activeStates().entrySet().stream()
-        .map(entry -> InstanceState.create(entry.getKey(), entry.getValue()))
-        .collect(toList());
-
-    return activeStates.stream()
-        .map(state -> state.runState().data().trigger())
-        .filter(Optional::isPresent)
-        .map(Optional::get)
-        .filter(TriggerUtil::isBackfill)
-        .collect(groupingBy(
-            TriggerUtil::triggerId,
-            HashMap::new,
-            counting()));
+    try {
+      // race conditions:
+      //   - the other scheduler instance has triggered partition, but hasn't committed the
+      //     transaction, and the workflow instance has finished very quickly
```

Comment:
Hm, how can this happen? The instance can't get submitted to GKE before transactional state transition goes through commit, can it?

Reply:
The actual triggering (state transition) happens synchronously but on a separated transaction with respect to the one created in this class (the latter handling the backfill entity only). This implementation is indeed a bit different from the one we discussed irl.

Reply:
Or maybe you mean the other scheduler instance has triggered the partition and transitioned through many states, in fact up until finishing it and becoming inactive (each of those is a committed transaction) and then this instance triggers it again...

Reply:
> Or maybe you mean the other scheduler instance has triggered the partition and transitioned through many states, in fact up until finishing it and becoming inactive (each of those is a committed transaction) and then this instance triggers it again...

Precisely. Very remote case nevertheless, and we would still not pollute `nextTrigger` anyway because that part of the transaction would fail.